extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,262 @@
+/* Copyright 2020 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/basic_auth/plugin.h"
+
+#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "extensions/common/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include "extensions/metadata_exchange/base64.h"
+
+#else
+
+#include "common/common/base64.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+PROXY_WASM_NULL_PLUGIN_REGISTRY;
+
+using Base64 = Envoy::Base64;
+
+#endif
+
+using ::nlohmann::json;
+using ::Wasm::Common::JsonArrayIterate;
+using ::Wasm::Common::JsonGetField;
+using ::Wasm::Common::JsonObjectIterate;
+using ::Wasm::Common::JsonValueAs;
+
+static RegisterContextFactory register_BasicAuth(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+namespace {
+void deniedNoBasicAuthData() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. No Basic "
+                    "Authentication information found.",
+                    "", {});
+}
+
+void deniedInvalidCredentials() {
+  sendLocalResponse(401,
+                    "Request denied by Basic Auth check. Invalid "
+                    "username and/or password",
+                    "", {});
+}
+
+FilterHeadersStatus credentialsCheck(
+    const PluginRootContext::BasicAuthConfigRule& rule,
+    std::string_view authorization_header) {
+  // Check if the Basic auth header starts with "Basic "
+  std::string_view authorization_header_strip =
+      absl::StripPrefix(authorization_header, "Basic ");
+  // If authorization_header_strip is the string as the original string
+  // it means there was no "Basic " prefix found.
+  if (authorization_header_strip == authorization_header) {
```

Comment:
nit: instead of checking equality here, I would suggest at the beginning of the function, Use StartWith check whether authorization header start with `Basic `